pvanalytics/tests/quality/test_data_shifts.py

```diff
@@ -0,0 +1,102 @@
+    df = pd.read_csv(test_file_1)
+    signal_no_index = df['value']
+    df.index = pd.to_datetime(df['timestamp'])
+    signal_datetime_index = df['value']
+    changepoint_date = df[df['label'] == 1].index[0]
+    df_weekly_resample = df.resample('W').median()['value']
+    return (signal_no_index, signal_datetime_index,
+            df_weekly_resample, changepoint_date)
+
+
+def requires_ruptures(test):
+    """Skip `test` if ruptures is not installed."""
+    try:
+        import ruptures  # noqa: F401
+        has_ruptures = True
+    except ImportError:
+        has_ruptures = False
+    return pytest.mark.skipif(
+        not has_ruptures, reason="requires ruptures")(test)
+
+
+@requires_ruptures
+def test_detect_data_shifts(generate_time_series):
+    """
+    Unit test that data shifts are correctly identified in the simulated time
+    series.
+    """
+    signal_no_index, signal_datetime_index, df_weekly_resample, \
+        changepoint_date = generate_time_series
+    # Test that an error is thrown when a Pandas series with no datetime
+    # index is passed
+    pytest.raises(TypeError, dt.detect_data_shifts, signal_no_index)
+    # Test that an error is thrown when an incorrect ruptures method is passed
+    pytest.raises(TypeError, dt.detect_data_shifts, signal_datetime_index,
+                  True, False, "Pelt")
+    # Test that an error is thrown when weekly data is passed
+    pytest.raises(ValueError, dt.detect_data_shifts, df_weekly_resample)
+    # Test that an error is thrown when an incorrect string is passed as the
+    # cost variable
+    pytest.raises(ValueError, dt.detect_data_shifts, signal_datetime_index,
+                  True, False, ruptures.Binseg, "none")
+    # Test that a data shift is successfully detected within 5 days of
+    # inserted changepoint
+    shift_index = dt.detect_data_shifts(time_series=signal_datetime_index)
+    shift_index_dates = list(shift_index[shift_index].index)
+    # Test that the column name is handled if a series with no name is passed
+    signal_unnamed = signal_datetime_index.rename(None)
+    shift_index_unnamed = dt.detect_data_shifts(signal_unnamed)
+    shift_index_unnamed_dates = list(
+        shift_index_unnamed[shift_index_unnamed].index)
+    # Run model with manually entered parameters
+    shift_index_param = dt.detect_data_shifts(signal_datetime_index, True,
+                                              False, ruptures.BottomUp, "rbf")
+    shift_index_param_dates = list(
+        shift_index_param[shift_index_param].index)
+    # Test that an Import error is thrown when ruptures is not available
+    _temp_ruptures = sys.modules['ruptures']
+    sys.modules['ruptures'] = None
+    assert requires_ruptures(None).mark.kwargs['reason'] == 'requires ruptures'
+    pytest.raises(ImportError, dt.detect_data_shifts, signal_datetime_index)
+    # Re-import ruptures
+    sys.modules['ruptures'] = _temp_ruptures
```

Comment:
Uff, I appreciate the effort for good test coverage, but I don't think perfect coverage is important enough to make it worth messing with `sys.modules` here.  The equivalent in `quality.time` isn't covered, so there's precedent: https://coveralls.io/builds/49454826/source?filename=pvanalytics%2Fquality%2Ftime.py#L148

Reply:
Removed these tests and piped over requires_ruptures to conftest in https://github.com/pvlib/pvanalytics/pull/142/commits/1f6c786e7511cd63d108759a78fdde3ff9217a1f